### ISSUE TITLE
Solve MultipleConstraintsOnSmallInputTest.cpp with libFuzzer

### ISSUE THREAD

**Issue by Dor1s** [google/sanitizers#964]
... or satisfy as many constraints as possible, at least: https://github.com/llvm-mirror/compiler-rt/blob/master/test/fuzzer/MultipleConstraintsOnSmallInputTest.cpp

**Comment by Dor1s** [comment]
As mentioned in the [initial CL](https://reviews.llvm.org/D46924), libFuzzer was able to satisfy 100 constraints in less than hour and then got stuck.

Another thing I've just tried was to add `-only_ascii=1`, that helped to satisfy 126 constraints.

My next step would be to try a new mutation. Hopefully, not too specific for that task, but let's discuss it later, when and if it actually helps.

**Comment by kcc** [comment]
I'm afraid that this test is the worst case of greedy fuzzers being stuck in local max. 
We probably need more search backtracking (like randomly discarding half of the corpus and it's coverage). This is what ClusterFuzz does, BTW. 
Maybe add this test to oss-fuzz and see if ClusterFuzz's backtracking helps?

**Comment by Dor1s** [comment]
Yeah, that's a nice idea!

**Comment by Dor1s** [comment]
Uploaded https://github.com/google/oss-fuzz/pull/1485

**Comment by kcc** [comment]
If I read the data correctly, oss-fuzz has satisfied 196 constraints in one day.

**Comment by Dor1s** [comment]
Indeed! Actually, it has satisfied 195 and now is trying to break through 196th, but not a big deal. Nice progress, I'd say. You're definitely right about getting stuck in local max. I couldn't make it that far running locally. Attached coverage report: [cov.html.txt](https://github.com/google/sanitizers/files/2076572/cov.html.txt)

**Comment by Dor1s** [comment]
And here we go: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=8770

**Comment by dvyukov** [comment]
What's there?
```
You do not have permission to view the requested page. 
Reason: User is not allowed to view this issue
```

**Comment by Dor1s** [comment]
CC'ed you on the bug + removed restrictions. There is a crash input found for https://github.com/llvm-mirror/compiler-rt/blob/master/test/fuzzer/MultipleConstraintsOnSmallInputTest.cpp#L4

Basically, Kostya was right, CF's set of strategies (corpus subset of different sizes and other things) seems to be helpful for that type of problem. I've been running that target locally non-stop and couldn't get more than 125 constraints covered.

**Comment by kcc** [comment]
backtracking FTW

**Comment by kcc** [comment]
Any news here?
Did AFL solve this?
Do we have backtracking for AFL? 

Can we get more puzzles tried this way? 
E.g. I'd like to try OnlySomeBytesTest

**Comment by Dor1s** [comment]
> Any news here?
> Did AFL solve this?
AFL hasn't solved it (yet?).

> Do we have backtracking for AFL?
I think we do, but will double check (cc @jonathanmetzman) + share the coverage reached by AFL. The last time I've checked it was a month ago: https://github.com/google/oss-fuzz/pull/1495#issuecomment-396286892

> Can we get more puzzles tried this way?
> E.g. I'd like to try OnlySomeBytesTest

Sure, will add that if you find it interesting!

**Comment by Dor1s** [comment]
AFL still can't pass the condition number 113. 
[multiple_constraints_on_small_input_afl_fuzzer_coverage.html.txt](https://github.com/google/sanitizers/files/2193815/multiple_constraints_on_small_input_afl_fuzzer_coverage.html.txt)

Uploaded https://github.com/google/oss-fuzz/pull/1617
